apps/dashr-ncbi-explorer/app.R

```diff
@@ -0,0 +1,654 @@
+  params = list(
+    input(id = 'genbank-sequence', property = 'data'),
+    input(id ='cpg-options', property = 'value')
+  ),
+  
+  update_sequence <- function(data, value) {
+    if (value != 1) {
+      dna_data <- readDNAStringSet("data/random_fasta.FASTA", "fasta")
+      return(as.character(dna_data[[value]]))
+    }
+    else{
+      return(data[["sequence"]])
+    }
+  }
+)
+
+# Callback for Alignment-Viewer Output
+
+app$callback(
+  output(id = 'alignment-container', property = 'children'),
+  params = list(
+    input(id = 'submit-button-2', property = 'n_clicks'),
+    input(id = 'dataframe-store', property = "data"),
+    input(id = "table", property = "active_cell"),
+    state(id = 'genbank-input', property = 'value')
+  ),
+  
+  update_fasta <- function(n_clicks, search_data, cell, accession_id) {
+
+    accession_id <- gsub("\\s", "", accession_id)
+    
+    accession_id <- gsub('"', "", accession_id)
+    
+    accession_id <- strsplit(accession_id, ",")
+    
+    accession_id <- unlist(accession_id)
+    
+    ctx = app$callback_context()
+    
+    if (ctx$triggered$prop_id == "submit-button-2.n_clicks") {
+      bio_file <- read.GenBank(access.nb = accession_id)
+      write.dna(bio_file, file ="data/random_fasta.FASTA", format = "fasta", append = TRUE, nbcol = 6, colsep = "", colw = 10)
+      fasta_file <- toupper(read_file("data/random_fasta.FASTA"))
+      alignment_chart <- dashbioAlignmentChart(id = 'alignment-chart', data = fasta_file, height = 750, width = 1150)
+    }
+    
+    else if (!is.null(search_data[[1]]) && !is.null(cell[[1]])) {
+      search_data  <-  as.data.frame(matrix(unlist(search_data), nrow=length(unlist(search_data[1]))), stringsAsFactors = FALSE)
+      accession_id <- search_data[1, cell$row + 1]
+      accession_id <- gsub('"', "", accession_id)
+      accession_id <- gsub(" ", "", accession_id)
+      
+      bio_file <- read.GenBank(access.nb = accession_id)
+      write.dna(bio_file, file ="data/random_fasta.FASTA", format = "fasta", append = TRUE, nbcol = 6, colsep = "", colw = 10)
+      fasta_file <- toupper(read_file("data/random_fasta.FASTA"))
+      alignment_chart <- dashbioAlignmentChart(id = 'alignment-chart', data = fasta_file, height = 750, width = 1150)
+    }
+    
+    
+    else if (n_clicks < 1) {
+      alignment_chart <- dashbioAlignmentChart(
+        id = 'alignment-chart',
+        data = read_file("data/alignment_viewer_p53_clustalo.FASTA"),
+        height = 750,
+        width = 1150,
+        opacity = 0.5
+      )
+    }
+    return(alignment_chart)
+  }
+)
+
+#Callbacks for database search output.
+
+app$callback(
+  output(id = "search-output", property = "children"),
+  params = list(
+    input(id = "search-button", property = "n_clicks"),
+    state(id = "search-input", property = "value")
+  ),
+  
+  update_search_results <- function(n_clicks, search_term) {
+    if (n_clicks > 0) {
+      search_term <- gsub('"', "", search_term)
```

Comment:
'      search_term <- gsub('"', "", search_term)

      search <- entrez_search(db = "nuccore", term = search_term, retmax=10)

      search_seqs <- entrez_fetch(db = "nuccore", id = search$ids, rettype = "fasta")

      red <- unlist(str_extract_all(search_seqs, ">.*\\n"))

      titles_vector <- c()

      for (i in red) {
        title <- str_extract_all(i, ">.*?[:blank:]")
        title <- substring(title, 2)
        titles_vector <- c(titles_vector, title)
      }

      sequence_titles <- as.list(unlist(str_extract_all(search_seqs, ">.*\\n")))

      sequence_dataframe <- do.call(rbind.data.frame, sequence_titles)

      names(sequence_dataframe)[1] <- "Dataset Accession ID and Description"

      sequence_dataframe$Accession_ID <- titles_vector

      sequence_dataframe <- sequence_dataframe[, c(2,1)]' 


this could be a function also (with title as a parameter if it changes infuture), anything long referenced multiple time could be placed in a helpers file to avoid repeating and making code base longer!

Reply:
Great suggestion, will implement. Thank you.